kernel/aster-nix/src/syscall/pwritev.rs

```diff
@@ -0,0 +1,135 @@
+    let flags = match RWFFlag::from_bits(flags) {
+        Some(flags) => flags,
+        None => return_errno_with_message!(Errno::EINVAL, "invalid flags"),
+    };
+    let res = if offset == -1 {
+        do_sys_writev(fd, io_vec_ptr, io_vec_count)?
+    } else {
+        do_sys_pwritev(fd, io_vec_ptr, io_vec_count, offset, flags)?
+    };
+    Ok(SyscallReturn::Return(res as _))
+}
+
+fn do_sys_pwritev(
+    fd: FileDesc,
+    io_vec_ptr: Vaddr,
+    io_vec_count: usize,
+    offset: i64,
+    _flags: RWFFlag,
+) -> Result<usize> {
+    // TODO: Implement flags support
+    debug!(
+        "fd = {}, io_vec_ptr = 0x{:x}, io_vec_counter = 0x{:x}, offset = 0x{:x}",
+        fd, io_vec_ptr, io_vec_count, offset
+    );
+    if offset < 0 {
+        return_errno_with_message!(Errno::EINVAL, "offset cannot be negative");
+    }
+    let file = {
+        let current = current!();
+        let filetable = current.file_table().lock();
+        filetable.get_file(fd)?.clone()
+    };
+    // TODO: Check (f.file->f_mode & FMODE_PREAD); We don't have f_mode in our FileLike trait
+    if io_vec_count == 0 {
+        return Ok(0);
+    }
+    if offset.checked_add(io_vec_count as i64).is_none() {
+        return_errno_with_message!(Errno::EINVAL, "offset + user_buf_len overflow");
+    }
+
+    let mut total_len = 0;
+    let mut cur_offset = offset as usize;
+    for i in 0..io_vec_count {
+        let io_vec = read_val_from_user::<IoVec>(io_vec_ptr + i * core::mem::size_of::<IoVec>())?;
+        if io_vec.base == 0 || io_vec.len == 0 {
+            continue;
+        }
+        let buffer = {
+            let base = io_vec.base;
+            let len = io_vec.len;
+            let mut buffer = vec![0u8; len];
+            read_bytes_from_user(base, &mut buffer)?;
+            buffer
+        };
+        let write_len = file.write_at(cur_offset, &buffer)?;
+        total_len += write_len;
+        cur_offset += write_len;
+    }
+    Ok(total_len)
+}
+
+fn do_sys_writev(fd: FileDesc, io_vec_ptr: Vaddr, io_vec_count: usize) -> Result<usize> {
+    debug!(
+        "fd = {}, io_vec_ptr = 0x{:x}, io_vec_counter = 0x{:x}",
+        fd, io_vec_ptr, io_vec_count
+    );
+    let file = {
+        let current = current!();
+        let filetable = current.file_table().lock();
+        filetable.get_file(fd)?.clone()
+    };
+    let mut total_len = 0;
+    for i in 0..io_vec_count {
+        let io_vec = read_val_from_user::<IoVec>(io_vec_ptr + i * core::mem::size_of::<IoVec>())?;
+        if io_vec.base == 0 || io_vec.len == 0 {
+            continue;
+        }
+        let buffer = {
+            let base = io_vec.base;
+            let len = io_vec.len;
+            let mut buffer = vec![0u8; len];
+            read_bytes_from_user(base, &mut buffer)?;
+            buffer
+        };
+        let write_len = file.write(&buffer)?;
+        total_len += write_len;
+    }
+    Ok(total_len)
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+struct IoVec {
```

Comment:
The definition of `IoVec` is redundant in multiple files. Please define the struct in `src/util/io_vec.rs`, then all syscalls can use the only definition.